torch/lib/c10d/test/TCPStoreTest.cpp

```diff
@@ -136,10 +151,221 @@ void testHelper(const std::string& prefix = "") {
+void testKeyChangeHelper(
+    c10d::Store& store,
+    std::string key,
+    const c10::optional<std::string>& expectedOldValue,
+    const c10::optional<std::string>& expectedNewValue) {
+  std::exception_ptr eptr = nullptr;
+  std::promise<bool> callbackPromise;
+
+  // Test the correctness of new_value and old_value
+  c10d::StoreCallbackFunction callback =
+      [expectedOldValue, expectedNewValue, &callbackPromise, &eptr, &key](
+          c10::optional<std::string> oldValue,
+          c10::optional<std::string> newValue) {
+        try {
+          EXPECT_EQ(
+              expectedOldValue.value_or("NONE"), oldValue.value_or("NONE"));
+          EXPECT_EQ(
+              expectedNewValue.value_or("NONE"), newValue.value_or("NONE"));
+        } catch (...) {
+          eptr = std::current_exception();
+        }
+        callbackPromise.set_value(true);
+      };
+  store.watchKey(key, callback);
+
+  // Perform the specified update according to key
+  if (key == "testEmptyKeyValue") {
+    c10d::test::set(store, key, expectedNewValue.value());
+  } else if (key == "testRegularKeyValue") {
+    c10d::test::set(store, key, expectedNewValue.value());
+  } else if (key == "testWatchKeyCreate") {
+    c10d::test::set(store, key, expectedNewValue.value());
+  } else if (key == "testWatchKeyAdd") {
+    store.add(key, std::stoi(expectedNewValue.value()));
+  } else if (key == "testWatchKeyDelete") {
+    store.deleteKey(key);
+  }
+
+  // Test that the callback is fired and the expected values are correct
+  std::future<bool> callbackFuture = callbackPromise.get_future();
+  std::chrono::milliseconds span(kStoreCallbackTimeoutMillis);
+  if (callbackFuture.wait_for(span) == std::future_status::timeout)
+    throw std::runtime_error("Callback execution timed out.");
+
+  // Any exceptions raised from asserts should be rethrown
+  if (eptr)
+    std::rethrow_exception(eptr);
+}
+
+TEST(TCPStoreTest, testKeyEmptyUpdate) {
+  auto store = _createServer();
+
+  std::string key = "testEmptyKeyValue";
+  c10d::test::set(*store, key, "");
+  store->get(key);
+  testKeyChangeHelper(*store, key, "", "2");
+}
+
+TEST(TCPStoreTest, testKeyUpdate) {
+  auto store = _createServer();
+
+  std::string key = "testRegularKeyValue";
+  c10d::test::set(*store, key, "1");
+  store->get(key);
+  testKeyChangeHelper(*store, key, "1", "2");
+}
+
+TEST(TCPStoreTest, testKeyCreate) {
+  auto store = _createServer();
+
+  std::string key = "testWatchKeyCreate";
+  testKeyChangeHelper(*store, key, c10::nullopt, "2");
+}
+
+TEST(TCPStoreTest, testKeyAdd) {
+  auto store = _createServer();
+
+  std::string key = "testWatchKeyAdd";
+  testKeyChangeHelper(*store, key, c10::nullopt, "2");
+}
+
+TEST(TCPStoreTest, testKeyDelete) {
+  auto store = _createServer();
+
+  std::string key = "testWatchKeyDelete";
+  c10d::test::set(*store, key, "1");
+  store->get(key);
+  testKeyChangeHelper(*store, key, "1", c10::nullopt);
+}
```

Comment:
Looks like these tests are all testing server stores callbacks? Do we have tests covering the case where callbacks are installed on client stores?

Reply:
Yes, these are all just for one process (acts as both server and client to itself) to simplify things since I wanted to test the correctness of the returned callback arguments for each different operation. I will update the multiple process test to add callbacks to server along with all clients as you mentioned in the previous comment!